src/lib.rs

```diff
@@ -142,98 +64,110 @@ impl Client {
-        &self,
-        path: &str,
-        json: &T,
-    ) -> error::Result<R> {
+    pub(crate) fn fetch_data<T, Q>(&self, path: &str, query: &Q) -> Future<Data<T>>
+    where
+        T: 'static + DeserializeOwned + std::marker::Send,
+        Q: Serialize + ?Sized,
+    {
         let request_url = format!("{}{}", self.base_url, path);
-        let response = self
-            .client
-            .post(&request_url)
-            .json(json)
-            .send()?
-            .error_for_status()?;
-        let result: Data<R> = response.json()?;
-        Ok(result.data)
-    }
-
-    /// Get wallet information for a given address
-    pub fn get_account(&self, address: &str) -> error::Result<Account> {
-        self.fetch::<Account>(&format!("/accounts/{}", address))
-    }
-
-    /// Get the current block height
-    pub fn get_height(&self) -> error::Result<u64> {
-        let result = self.fetch::<Height>("/blocks/height")?;
-        Ok(result.height)
-    }
-
-    /// Get hotspots for a given wallet address
-    pub fn get_hotspots(&self, address: &str) -> error::Result<Vec<Hotspot>> {
-        self.fetch::<Vec<Hotspot>>(&format!("/accounts/{}/hotspots", address))
+        self.client
+            .get(&request_url)
+            .query(query)
+            .send()
+            .map_err(error::Error::from)
+            .and_then(|response| match response.error_for_status() {
+                Ok(result) => {
+                    let data: Future<Data<T>> = result.json().map_err(error::Error::from).boxed();
+                    data
+                }
+                Err(e) => future::err(error::Error::from(e)).boxed(),
+            })
+            .boxed()
     }
 
-    /// Get details for a given hotspot address
-    pub fn get_hotspot(&self, address: &str) -> error::Result<Hotspot> {
-        self.fetch::<Hotspot>(&format!("/hotspots/{}", address))
+    pub(crate) fn fetch_stream<E, Q>(self, path: &str, query: &Q) -> Stream<E>
+    where
+        E: 'static + DeserializeOwned + std::marker::Send,
+        Q: Serialize + ?Sized,
+    {
+        let path = path.to_string();
+        let client = self;
+        client
+            .fetch_data::<Vec<E>, _>(&path, query)
+            .map_ok(move |mut data| {
+                data.data.reverse();
+                stream::try_unfold(
+                    (data, client, path),
+                    |(mut data, client, path)| async move {
+                        match data.data.pop() {
+                            Some(entry) => Ok(Some((entry, (data, client, path)))),
+                            None => match data.cursor {
+                                Some(cursor) => {
+                                    let mut data = client
+                                        .fetch_data::<Vec<E>, _>(&path, &[("cursor", &cursor)])
+                                        .await?;
+                                    data.data.reverse();
```

Comment:
What's the reverse for? If you fetch a stream of `blocks` they start from highest and go to lower thanks to the reverse before popping?

Reply:
exactly